cmd/util/ledger/migrations/cadence_value_validation.go

```diff
@@ -0,0 +1,599 @@
+}
+
+func cadenceArrayValueEqual(
+	vInterpreter *interpreter.Interpreter,
+	v *interpreter.ArrayValue,
+	otherInterpreter *interpreter.Interpreter,
+	other interpreter.Value,
+) *validationError {
+	otherArray, ok := other.(*interpreter.ArrayValue)
+	if !ok {
+		return newValidationErrorf("types differ: %T != %T", v, other)
+	}
+
+	count := v.Count()
+	if count != otherArray.Count() {
+		return newValidationErrorf("array counts differ: %d != %d", count, otherArray.Count())
+	}
+
+	if v.Type == nil {
+		if otherArray.Type != nil {
+			return newValidationErrorf("array types differ: nil != %s", otherArray.Type)
+		}
+	} else { // v.Type != nil
+		if otherArray.Type == nil {
+			return newValidationErrorf("array types differ: %s != nil", v.Type)
+		} else if !v.Type.Equal(otherArray.Type) {
+			return newValidationErrorf("array types differ: %s != %s", v.Type, otherArray.Type)
+		}
+	}
+
+	for i := 0; i < count; i++ {
+		element := v.Get(vInterpreter, interpreter.EmptyLocationRange, i)
+		otherElement := otherArray.Get(otherInterpreter, interpreter.EmptyLocationRange, i)
+
+		err := cadenceValueEqual(vInterpreter, element, otherInterpreter, otherElement)
+		if err != nil {
+			err.addTrace(fmt.Sprintf("(%s[%d])", v.Type, i))
+			return err
+		}
+	}
+
+	return nil
+}
+
+func cadenceCompositeValueEqual(
+	vInterpreter *interpreter.Interpreter,
+	v *interpreter.CompositeValue,
+	otherInterpreter *interpreter.Interpreter,
+	other interpreter.Value,
+) *validationError {
+	otherComposite, ok := other.(*interpreter.CompositeValue)
+	if !ok {
+		return newValidationErrorf("types differ: %T != %T", v, other)
+	}
+
+	if !v.StaticType(vInterpreter).Equal(otherComposite.StaticType(otherInterpreter)) {
+		return newValidationErrorf(
+			"composite types differ: %s != %s",
+			v.StaticType(vInterpreter),
+			otherComposite.StaticType(otherInterpreter),
+		)
+	}
+
+	if v.Kind != otherComposite.Kind {
+		return newValidationErrorf(
+			"composite kinds differ: %d != %d",
+			v.Kind,
+			otherComposite.Kind,
+		)
+	}
+
+	var err *validationError
+	vFieldNames := make([]string, 0, 10) // v's field names
+	v.ForEachField(nopMemoryGauge, func(fieldName string, fieldValue interpreter.Value) bool {
+		otherFieldValue := otherComposite.GetField(otherInterpreter, interpreter.EmptyLocationRange, fieldName)
+
+		err = cadenceValueEqual(vInterpreter, fieldValue, otherInterpreter, otherFieldValue)
+		if err != nil {
+			err.addTrace(fmt.Sprintf("(%s.%s)", v.TypeID(), fieldName))
+			return false
+		}
+
+		vFieldNames = append(vFieldNames, fieldName)
+		return true
+	})
+
+	// TODO: Use CompositeValue.FieldCount() from Cadence after it is merged and available.
+	otherFieldNames := make([]string, 0, len(vFieldNames)) // otherComposite's field names
+	otherComposite.ForEachField(nopMemoryGauge, func(fieldName string, _ interpreter.Value) bool {
+		otherFieldNames = append(otherFieldNames, fieldName)
+		return true
+	})
+
+	if len(vFieldNames) != len(otherFieldNames) {
```

Comment:
Do we need to compare if they contain the same items?

If so, what about building a map, and remove each field from the map when iterating through `otherComposite`, and verify if there is any field left.

Reply:
> Do we need to compare if they contain the same items?

Not needed here because:
- we compare number of fields for both composite values
- we compare fields in 2 composite values by iterating over fields of 1st composite value and comparing field values (of matching field name) in 2nd composite value
- if composite values have different fields, we would fail to get field value by the same field name in the previous step.